packages/lodestar/src/network/attestation-service/service.ts

```diff
@@ -0,0 +1,384 @@
+          this.subscribeToRandomSubnet(currentFork);
+        }
+      }
+    }
+    this.knownValidators.set(validatorIndex, currentSlot + LAST_SEEN_VALIDATOR_TIMEOUT);
+  }
+
+  /**
+   * Subscribe to long-lived random subnets and update the local ENR bitfield.
+   */
+  private subscribeToRandomSubnet(fork: ForkName): void {
+    const currentSlot = this.chain.clock.currentSlot;
+    const allSubnets = Array.from({length: ATTESTATION_SUBNET_COUNT}, (_, i) => i);
+    const activeSubnets = this.randomSubnets.getActive(currentSlot);
+    const availableSubnets = allSubnets.filter((subnet) => !activeSubnets.includes(subnet));
+    const toSubscribeSubnet = availableSubnets[randBetween(0, availableSubnets.length)];
+    // the heartbeat will help connect to respective peers
+    this.randomSubnets.request({
+      subnetId: toSubscribeSubnet,
+      toSlot: getSubscriptionSlotForRandomSubnet(this.config, currentSlot),
+    });
+    if (!this.subscribedCommitteeSubnets.getActive(currentSlot).includes(toSubscribeSubnet)) {
+      // subscribe to topic
+      this.gossip.subscribeTopic({type: GossipType.beacon_attestation, fork, subnet: toSubscribeSubnet});
+    }
+    // Update ENR
+    const attnets = this.metadata.attnets;
+    if (!attnets[toSubscribeSubnet]) {
+      attnets[toSubscribeSubnet] = true;
+      this.metadata.attnets = attnets;
+    }
+  }
+
+  /**
+   * Duplicate randomSubnets to nextForkRandomSubnets to make sure same ENR advertisement
+   * Subscribe to respective gossip topics with next fork.
+   * No need to update ENR
+   * @param nextFork
+   */
+  private subscribeToNextForkRandomSubnets(nextFork: ForkName): void {
+    const currentSlot = this.chain.clock.currentSlot;
+    const activeSubnets = this.randomSubnets.getActive(currentSlot);
+    for (const subnet of activeSubnets) {
+      this.nextForkRandomSubnets!.request({
+        subnetId: subnet,
+        toSlot: getSubscriptionSlotForRandomSubnet(this.config, currentSlot),
+      });
+      this.gossip.subscribeTopic({type: GossipType.beacon_attestation, fork: nextFork, subnet});
+    }
+  }
+
+  /**
+   * Run per slot.
+   */
+  private onSlot = (slot: Slot): void => {
+    try {
+      const fork = this.chain.getForkName();
+      if (fork !== this.randomSubnets.getForkName() && fork === this.nextForkRandomSubnets?.getForkName()) {
+        this.transitionToNextFork();
+      }
+      this.handleRandomSubnetExpiry(slot);
+      this.handleKnownValidatorExpiry(slot);
+      this.handleUnsubscriptions(slot);
+    } catch (e) {
+      this.logger.error("Error running AttestationService at slot", {slot}, e);
+    }
+  };
+
+  private transitionToNextFork(): void {
+    const fork = this.chain.getForkName();
+    assert.true(this.nextForkRandomSubnets !== undefined, "No next fork random subnets subscriptions");
+    assert.equal(
+      fork,
+      this.nextForkRandomSubnets?.getForkName(),
+      "Chain fork name is not same to next fork random subnet"
+    );
```

Comment:
If this conditions throw they will interrupt the flow of onSlot. We may not want that

Reply:
this is just like a precondition before running this function to make it easy to understand, we had a check in `onSlot` already so it should be safe.
It's ok to remove the `assert()` but I found it's easier to understand that way.